### PR TITLE
Add function utils

### PR DESCRIPTION
**Pull request by Flanker32**
Add function utils

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+    public static Module getFunctionModuleByFilePath(Project project, String moduleFilePath){
+        Module[] modules = listFunctionModules(project);
+        return Arrays.stream(modules)
+                .filter(module -> StringUtils.equals(moduleFilePath, module.getModuleFilePath()))
+                .findFirst().orElse(null);
+    }
+
+    public static boolean isFunctionProject(Project project) {
+        List<Library> libraries = new ArrayList<>();
+        OrderEnumerator.orderEntries(project).forEachLibrary(library -> libraries.add(library));
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: OrderEnumerator.orderEntries(module).productionOnly()

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+        return ListUtils.indexOf(libraries,
+                library -> StringUtils.contains(library.getName(), FUNCTION_JAVA_LIBRARY_ARTIFACT_ID)) >= 0;
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: we can combine two lamdas together to terminate stream when lib is found

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/JarUtils.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+public class JarUtils {
+
+    private static final int BUFFER_SIZE = 4096;
+
+    public static Path buildJarFileToStagingPath(String stagingFolder, Module module) {
+        String moduleName = module.getName();
+        Manifest manifest = new Manifest();
+        manifest.getMainAttributes().put(Attributes.Name.MANIFEST_VERSION, "1.0");
+        manifest.getMainAttributes().put(new Attributes.Name("Created-By"), "Azure Intellj Plugin");
+
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/JarUtils.java]: remove   Manifest manifest = new Manifest();
        manifest.getMainAttributes().put(Attributes.Name.MANIFEST_VERSION, "1.0");
        manifest.getMainAttributes().put(new Attributes.Name("Created-By"), "Azure Intellj Plugin");

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/JarUtils.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        String path = CompilerPaths.getModuleOutputPath(module, false);
+
+        final Path outputFile = Paths.get(stagingFolder, moduleName + ".jar");
+        try (final ZipOutputStream outputZip = getZipOutputStream(outputFile, true)) {
+            zipDirectory(new File(path), "", outputZip);
+            addManifest(outputZip);
+        } catch (IOException | AzureExecutionException e) {
+            e.printStackTrace();
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/JarUtils.java]: throws IOException, AzureExecutionException

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/JarUtils.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+    private static ZipOutputStream getZipOutputStream(Path outputFile, boolean override)
+            throws AzureExecutionException, IOException {
+
+        if (Files.exists(outputFile, LinkOption.NOFOLLOW_LINKS)) {
+            if (override) {
+                try {
+                    Files.delete(outputFile);
+                } catch (IOException exception) {
+                    throw new IllegalStateException("Removing \"" + outputFile + "\" failed. Not writing out anything.",
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/JarUtils.java]: IllegalStateException -> AzureExecutionException

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+                binding.setAttribute(name, result.toArray(new String[0]));
+            } else {
+                String valueText = getAnnotationValueAsString(value);
+                binding.setAttribute(name, valueText);
+            }
+        }
+        return binding;
+    }
+
+    public static String getEnumFieldString(final String className, final String fieldName)
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: public -> private

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+                }
+            });
+            // Replace empty connection string
+
+        } else {
+            System.out.println("No StorageAccount annotation found.");
+        }
+    }
+
+    public static String getAnnotationValueAsString2(JvmAnnotationAttributeValue value) throws AzureExecutionException {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: extract and rename into an helper

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+                PsiReferenceExpression referenceExpression = (PsiReferenceExpression) value;
+                Object resolved = referenceExpression.resolve();
+                if (resolved instanceof PsiEnumConstant) {
+                    final PsiEnumConstant enumConstant = (PsiEnumConstant) resolved;
+                    final PsiClass enumClass = enumConstant.getContainingClass();
+                    if (enumClass != null) {
+                        try {
+                            return getEnumFieldString(enumClass.getQualifiedName(), enumConstant.getName());
+                        } catch (ClassNotFoundException e) {
+                            e.printStackTrace();
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: wrapper with AzureExecutionException

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java (hunk 7 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: getAnnotationValueAsString2 => getAnnotationValueString

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+        PsiClass functionNameClass = JavaPsiFacade.getInstance(module.getProject())
+                .findClass(AZURE_FUNCTION_ANNOTATION_CLASS, GlobalSearchScope.moduleWithLibrariesScope(module));
+        List<PsiMethod> methods = new ArrayList<>(AnnotatedElementsSearch
+                .searchPsiMethods(functionNameClass, GlobalSearchScope.allScope(project)).findAll());
+        return methods.toArray(new PsiMethod[0]);
+    }
+
+    public static final Path getDefaultHostJson(Project project) {
+        File projectHost = new File(project.getBasePath(), "host.json");
+        return projectHost.exists() ? projectHost.toPath() : createTempleHostJson();
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: don't create temp file

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+
+    private static String stripExtraCharacters(String fileName) {
+        // TODO-dp this is not robust enough (eliminated !/ at the end of the jar)
+        if (fileName.endsWith("!/")) {
+            fileName = fileName.substring(0, fileName.length() - 2);
+        }
+        return fileName;
+    }
+
+    public static Map<String, FunctionConfiguration> generateConfigurations(final PsiMethod[] methods)
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: private static

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java (hunk 1 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: Fixed

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/AnnotationHelper.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        } else if (value instanceof PsiLiteral) {
+            return Objects.toString(((PsiLiteral) value).getValue(), null);
+        }
+        throw new AzureExecutionException("Cannot get annotation value of type : " + value.getClass().getName());
+    }
+
+    public static String getPsiAnnotationMemberValue(PsiAnnotationMemberValue value) throws AzureExecutionException {
+        if (value instanceof PsiExpression) {
+            if (value instanceof PsiReferenceExpression) {
+                final PsiReferenceExpression referenceExpression = (PsiReferenceExpression) value;
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/AnnotationHelper.java]: dup code as above method,

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/AnnotationHelper.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+                final Object resolved = referenceExpression.resolve();
+                if (resolved instanceof PsiEnumConstant) {
+                    final PsiEnumConstant enumConstant = (PsiEnumConstant) resolved;
+                    final PsiClass enumClass = enumConstant.getContainingClass();
+                    if (enumClass != null) {
+                        try {
+                            return getEnumFieldString(enumClass.getQualifiedName(), enumConstant.getName());
+                        } catch (ClassNotFoundException e) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/AnnotationHelper.java]: handle the exception with AzureExecutionException

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+        final Module[] modules = listFunctionModules(project);
+        return Arrays.stream(modules)
+                .filter(module -> StringUtils.equals(moduleFilePath, module.getModuleFilePath()))
+                .findFirst().orElse(null);
+    }
+
+    public static boolean isFunctionProject(Project project) {
+        final List<Library> libraries = new ArrayList<>();
+        OrderEnumerator.orderEntries(project).productionOnly().forEachLibrary(library -> libraries.add(library));
+        return ListUtils.indexOf(libraries, library ->
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: move StringUtils.contains(library.getName(), FUNCTION_JAVA_LIBRARY_ARTIFACT_ID
to forEachLibrary

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+                    if (file.exists()) {
+                        jarFiles.add(file);
+                    }
+                }
+            }
+            return true;
+        });
+
+        for (final File file : jarFiles) {
+            FileUtils.copyFileToDirectory(file, new File(stagingFolder.toFile(), "lib"));
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: extract  new File(stagingFolder.toFile(), "lib") to a var

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java [hunk 15]
```diff
@@ -0,0 +1,375 @@
...
+            return StringUtils.EMPTY;
+        }
+        final Project project = module.getProject();
+        final MavenProject mavenProject = MavenProjectsManager.getInstance(project).findProject(module);
+        final String stagingFolderName = mavenProject == null ? project.getName() : mavenProject.getProperties().getProperty("functionAppName");
+        return Paths.get(project.getBasePath(), "target", "azure-functions", stagingFolderName).toString();
+    }
+
+    public static String getFuncPath() throws IOException, InterruptedException {
+        final List<String> outputStrings = executeMultipLineOutput(
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: use new version of finder:
https://github.com/microsoft/azure-gradle-plugins/blob/master/azure-functions-gradle-plugin/src/main/java/com/microsoft/azure/plugin/functions/gradle/util/FunctionCliResolver.java

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java [hunk 16]
```diff
@@ -0,0 +1,375 @@
...
+        }
+    }
+
+    private static List<Binding> parseAnnotations(JvmAnnotation[] annos) throws AzureExecutionException {
+        final List<Binding> bindings = new ArrayList<>();
+
+        for (final JvmAnnotation annotation : annos) {
+            final Binding binding = getBinding(annotation);
+            if (binding != null) {
+                System.out.println("Adding binding: " + binding.toString());
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: Log.debug

Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: Fixed

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java (hunk 15 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: Copied

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java (hunk 14 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: Done

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java (hunk 13 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/FunctionUtils.java]: Fixed

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/AnnotationHelper.java (hunk 12 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/AnnotationHelper.java]: Done

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/AnnotationHelper.java (hunk 11 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/core/AnnotationHelper.java]: Will fix in later PR

**Pull request merged**